src/context/authStore.tsx

```diff
@@ -41,24 +37,74 @@ export const AuthStoreContext = createContext<AuthStoreContext>({
+ * Assumes that no one uses v1 storage anymore as of Jul 2021.
+ *
+ * Versions:
+ *
+ * 1. (deprecated)
+ *    - Stores the corresponding values in AsyncStorage under the following keys: {@link SESSION_TOKEN_KEY}, {@link EXPIRY_KEY} and {@link ENDPOINT_KEY}
+ *    - Can only store one set of credentials
+ * 2. (deprecated)
+ *    - Stores a stringified {@link AuthCredentialsMap} in AsyncStorage under the key {@link AUTH_CREDENTIALS_STORE_KEY}
+ *    - Map from {{@link AuthCredentials.operatorToken}}{{@link AuthCredentials.endpoint}} to corresponding credentials
+ * 3. (current)
+ *    - Stores a stringified {@link AuthCredentialsMap} in SecureStorage under the key {@link AUTH_CREDENTIALS_STORE_KEY}
+ *    - Map from a hash of {{@link AuthCredentials.operatorToken}}{{@link AuthCredentials.endpoint}} to corresponding credentials
+ *    - Due to size limit of SecureStore, uses a bucketing method explained in {@link saveToStoreInBuckets}
+ */
 export const AuthStoreContextProvider: FunctionComponent<{
   shouldMigrate?: boolean; // Temporary toggle to disable migration when testing
 }> = ({ shouldMigrate = true, children }) => {
+  /**
+   * Flag to mark when all loading has been done and consumers can read from the data.
+   * This includes any migration actions needed.
+   */
   const [hasLoadedFromStore, setHasLoadedFromStore] = useState(false);
+  /**
+   * This flag marks when {@link authCredentials} is matching with the current state of
+   * the primary store. Any changes after this flag is true should trigger updates to the
+   * primary store.
+   *
+   * This flag should be set to true before any migration happens so that migrations will
+   * update the primary store.
+   *
+   * Documented in [#398](https://github.com/rationally-app/mobile-application/pull/398)
+   */
+  const [hasLoadedFromPrimaryStore, setHasLoadedFromPrimaryStore] = useState(
+    false
+  );
   const [authCredentials, setAuthCredentialsMap] = useState<AuthCredentialsMap>(
     {}
   );
   const prevAuthCredentials = usePrevious(authCredentials);
 
+  const [, setState] = useState();
+  /** Effect to update the store when there are changes to authCredentials */
   useEffect(() => {
-    if (hasLoadedFromStore) {
+    if (hasLoadedFromPrimaryStore) {
       const authCredentialsString = JSON.stringify(authCredentials);
       const prevAuthCredentialsString = JSON.stringify(prevAuthCredentials);
+      // do a top level check to see if there are any changes
       if (authCredentialsString !== prevAuthCredentialsString) {
-        AsyncStorage.setItem(AUTH_CREDENTIALS_STORE_KEY, authCredentialsString);
+        // if there are changes, do a check for each bucket
+        saveToStoreInBuckets(
+          AUTH_CREDENTIALS_STORE_KEY,
+          authCredentialsString,
+          prevAuthCredentialsString ?? ""
+        ).catch((reason) => {
+          Sentry.addBreadcrumb({
+            category: "authStore",
+            message: "save failed",
+          });
```

Comment:
Just wondering, is there a reason why you're adding a breadcrumb and not creating a sentry exception?

Reply:
decided to throw it up to the error boundary which will log as an exception
The alternative is to proceed without saving credentials in the storage so everything will be temporary. Is that preferred?

Reply:
decided to throw it up to the error boundary which will log as an exception
The alternative is to proceed without saving credentials in the storage so everything will be temporary. Is that preferred?